imi/src/main/java/org/motechproject/nms/imi/service/impl/CdrFileServiceImpl.java

```diff
@@ -744,4 +753,65 @@ public void verifyDetailFileChecksumAndCount(CdrFileNotificationRequest request)
         }
         return errors;
     }
+
+    @Override
+    @MotechListener(subjects = { CDR_CSR_CLEANUP_SUBJECT })
+    public void cleanOldCallRecords() {
+        LOGGER.info("cleanOldCallRecords() called");
+        int cdrDuration = MIN_CALL_DATA_RETENTION_DURATION_IN_DAYS;
+
+        try {
+            int durationFromConfig = Integer.parseInt(settingsFacade.getProperty(CDR_CSR_RETENTION_DURATION));
+            if (durationFromConfig < cdrDuration) {
+                LOGGER.debug(String.format("Discarding retention property from config since it is less than MIN: %d", cdrDuration));
+            } else {
+                cdrDuration = durationFromConfig;
+                LOGGER.debug(String.format("Using retention property from config: %d", cdrDuration));
+            }
+        } catch (NumberFormatException ne) {
+            LOGGER.debug(String.format("Unable to get property from config: %s", CDR_CSR_RETENTION_DURATION));
+        }
+
+        String logTemplate = "Found %d records in table %s";
+
+        LOGGER.debug(String.format(logTemplate, callDetailRecordDataService.count(), CDR_TABLE_NAME));
+        LOGGER.debug(String.format(logTemplate, callSummaryRecordDataService.count(), CSR_TABLE_NAME));
+
+        deleteRecords(cdrDuration, CDR_TABLE_NAME);
+        deleteRecords(cdrDuration, CSR_TABLE_NAME);
+
+        LOGGER.debug(String.format(logTemplate, callDetailRecordDataService.count(), CDR_TABLE_NAME));
+        LOGGER.debug(String.format(logTemplate, callSummaryRecordDataService.count(), CSR_TABLE_NAME));
+        callDetailRecordDataService.evictEntityCache(false);
+        callSummaryRecordDataService.evictEntityCache(false);
+    }
+
+    /**
+     * Helper to clean out the CDR table with the given retention policy
+     * @param retentionInDays min days to keep CDR for
+     * @param tableName name of the cdr or csr table
+     */
+    private void deleteRecords(final int retentionInDays, final String tableName) {
+        @SuppressWarnings("unchecked")
+        SqlQueryExecution<Long> queryExecution = new SqlQueryExecution<Long>() {
+
+            @Override
+            public String getSqlQuery() {
+                String query = String.format(
+                        "DELETE FROM %s where creationDate < now() - INTERVAL %d DAY", tableName, retentionInDays);
+                LOGGER.debug("SQL QUERY: {}", query);
+                return query;
+            }
+
+            @Override
+            public Long execute(Query query) {
+
+                return (Long) query.execute();
+            }
+        };
+
+        // FYI: doesn't matter what data service we use since it is just used as a vehicle to execute the custom query
```

Comment:
actually I think it does, because you want to empty the L1 cache. we should look into that. I wrote some code to empty the cache somewhere similar. let me see...